File: 0057-insert-interval/solution.java
class Solution {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        int n = intervals.length;

        int s1 = newInterval[0];//2
        int e1 = newInterval[1];//5
        ArrayList<int[]> list = new ArrayList<>();
        boolean inserted = false;// to control merging logic
        for(int i=0; i<n; i++) {
            int s2 = intervals[i][0];//1
            int e2 = intervals[i][1];//3

            if(e2 < s1) {
                // interval is before new interval here
                // no overlap , add original interval 
                list.add(new int[]{s2, e2});
            } else if(s2 <= e1) {
                // overlapping -> merge
                // update s1 and e1
                s1 = Math.min(s1, s2);
                e1 = Math.max(e1, e2);
            } else {
                // new interval is after current interval
                // add merged interval once
                // add the current interval
                // then stop merging
                if(!inserted) {
                    list.add(new int[]{s1, e1});// add merged interval once
                    inserted = true;
                }
                    list.add(new int[]{s2, e2}); // add the current interval
            }
        }
        if(!inserted) {
            list.add(new int[]{s1, e1});
        }

        int[][] ans = new int[list.size()][2];
        int ind = 0;
        for(int[] key: list) {
            ans[ind][0] = key[0];
            ans[ind][1] = key[1];
            ind++;
        }
        return ans;
    }
}

File: 0057-insert-interval/solution.swift
class Solution {
    func insert(_ intervals: [[Int]], _ newInterval: [Int]) -> [[Int]] {
        var ans = [[Int]]()
        var s1 = newInterval[0]//4
        var e1 = newInterval[1]//8
        var isMerged = false

        for interval in intervals {
            var s2 = interval[0]//6
            var e2 = interval[1]//7
            //---1--2--3--5--
            //----------4---8
            // case 1 - current interval before new interval
            // dont overlap
            if e2 < s1 {
                ans.append([s2,e2])//[[1,2], [3,8], [6,8], ]
            } 
            // overlap, update end interval
            else if s2 <= e1 {//4 <= 2, 4 <= 5, 4 <= 7
                // expand the new interval:
                // You should update the new interval, because that’s what you’re carrying forward
                s1 = min(s1, s2)
                e1 = max(e1, e2)
            } else {
                // new interval is after current interval
                // add merged interval once
                if !isMerged {
                    ans.append([s1,e1])//[[1,2], [3,8], [6,8], ]
                    // then stop merging from now on
                    isMerged = true
                }
                // append the current interval
                ans.append([s2,e2])//[[1,2], [3,8], [6,8], ]
            }
        }
        if !isMerged {
            ans.append([s1,e1])
        }
        return ans
    }
}
